### ISSUE TITLE
quad_form(x,P).is_dcp() is broken

### ISSUE THREAD

**Issue by bnaras** [cvxpy/cvxpy#1451]
**Describe the bug**
Gives answer `True` for an obviously `False` case.

**To Reproduce**
```
import cvxpy
import numpy as np
P = np.array([[1, 2], [2, 1]])
x = cvxpy.Variable(2)
q = cvxpy.quad_form(x, P)
q.is_dcp()
``` 

**Expected behavior**
Expecting an answer of `False`

**Output**
```
Python 3.8.8 (default, Apr 13 2021, 12:59:45) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import cvxpy as cp
>>> import numpy as np
>>> P = np.array([[1, 2], [2, 1]])
>>> x = cp.Variable(2)
>>> q = cp.quad_form(x, P)
>>> q.is_dcp()
True
>>> print(cp.__version__)
1.1.14
```

**Version**
 - OS: macOS 11.5.1
 - CVXPY Version:  1.1.14

**Additional context**
Version 1.1.1 seems to be the last version that yields `False`.

**Comment by akshayka** [comment]
Thanks for reporting.

I took a quick look. The logic for finding the top eigenvector of a constant looks wrong. For some reason, we're looking for an eigenvector whose eigenvalue is close to 0, which is obviously not what we want.

See https://github.com/cvxpy/cvxpy/blob/master/cvxpy/expressions/constants/constant.py#L239

**Comment by rileyjmurray** [comment]
@akshayka that code was introduced in response to people who had matrices that were (near) singular but still semidefinite (positive or negative). So the naming convention "top eig" in the context of "is_nsd()" just reflected an assumption that in the "typical" case we'd be looking for eigenvalues near zero. @bnaras's example shows that this assumption can actually affect correctness (rather than reliability for degenerate problems). The current implementation shows that ``[[1,2],[2,1]`` is reported as NSD because its largest eigenvalue *near zero* is negative. It's missing that the absolute largest eigenvalue is actually positive.

(I'll offer a solution in the next comment.)

**Comment by rileyjmurray** [comment]
One way forward is to use a basic method to bound all eigenvalues of the matrix ``P``. We could first apply [Gershgorin's circle theorem](https://en.wikipedia.org/wiki/Gershgorin_circle_theorem) to obtain intervals where the eigenvalues can be found. If all the the intervals consist of nonnegative numbers, then we can report that the matrix is PSD. If all intervals consist of nonpositive numbers, then we can report the matrix is NSD. If some interval contains zero, then we have to do some work.

First, we can set ``Q = P + lamb * I`` where ``lamb`` is a positive number chosen with consideration to the Gershgorin circle theorem so that ``Q`` is positive definite (say, minimum eigenvalue is at least 1). Then we can use the power method to compute the operator norm of the matrix ``Q`` (equivalent to ``nrm = np.linalg.norm(np.linalg.inv(Q), ord=2)``, but cheaper). I'm pretty sure the minimum eigenvalue of ``P`` will be equal to ``1/nrm - lamb``. I'll look into how we might call ``eigsh`` to compute the operator norm of ``np.linalg.inv(Q)`` without forming ``np.linalg.inv(Q)`` explicitly.

**Comment by SteveDiamond** [comment]
@rileyjmurray that method sounds pretty interesting, I'd be happy to see that as an option. Would you mind doing speed benchmarks though so we know how much this is buying us over the simple approach? Or is the gain in more accurate handling of near zero eigenvalues?

**Comment by rileyjmurray** [comment]
@SteveDiamond happy to. I'll try report back with *something* tomorrow. A full PR might need to wait until Saturday.

**Comment by rileyjmurray** [comment]
@SteveDiamond @akshayka I found a deeper cause of the problem.

The last time we made changes to how ``is_psd`` and ``is_nsd`` are computed for Constant objects, I added tests that used Parameter objects. The tests that I added meant to verify that Parameter objects that are declared as PSD or NSD could not accept indefinite values. The problem here is that the method that Parameter objects use to validate if an input matrix is PSD or NSD is separate from the method Constant objects use to decide if a matrix is PSD or NSD. Parameter objects do a full (dense) eigenvalue decomposition in ``leaf.py:Leaf.project(...)``, while Constants use the method that's been referenced here. So the first order of business is to have those tests use Constant objects rather than Parameter objects. We should also look at changing when Parameter objects report that they're PSD. Right now,  attributes for Parameter objects are decided only by their declared properties, rather than the property of the current value. This makes sense in some ways, but it also creates a problem where calling ``cp.Parameter(shape=(2,2), value=np.eye(2)).is_psd()`` returns ``False``.
